ipcache: Plumb daemon context through IPCache

joestringer opened the pull request:
Plumb the primary daemon context into the label injector logic and the
async prefix releaser in order to reduce the likelihood of issues where
this logic continues to execute beyond the lifetime of the agent.

Implement a Shutdown() function for the IPCache which shuts down the
goroutines associated with the IPcache and ensures they are completed
before returning. This way, other components should not continue to
cause the IPCache to asynchronously continue to execute work. This
should hopefully fix issues in unit tests where the IPCache defers some
work that interacts with the identityAllocator after it has been cleaned
up.

Found during unit testing, where the identity allocator would be closed
while the ipcache's garbage collection of identities was still running.

Suggested-by: Jussi Maki <jussi@isovalent.com>
Suggested-by: Aditi Ghag <aditi@cilium.io>

joestringer committed:
ipcache: Implement Shutdown

Implement a Shutdown() function for the IPCache which shuts down the
goroutines associated with the IPcache and ensures they are completed
before returning. This way, other components should not continue to
cause the IPCache to asynchronously continue to execute work. This
should hopefully fix issues in unit tests where the IPCache defers some
work that interacts with the identityAllocator after it has been cleaned
up.

Suggested-by: Aditi Ghag <aditi@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>

joamaki reviewed (approved):
LGTM!

aditighag commented on pkg/ipcache/gc.go:
```diff
@@ -41,15 +43,22 @@ func newAsyncPrefixReleaser(parent prefixReleaser, interval time.Duration) *asyn
+		result.Trigger.Shutdown()
+		close(result.closeChan)
+		result.Unlock()
+	}()
```
Is this goroutine simply running to check if the daemon context is done, and then trigger the shutdown in ipcache?
Have you looked into introducing a `Close`/`Shutdown` function in ipcache that'll be invoked during daemon shutdown - https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon.go#L1363? 
This approach has the benefit that we have some kind of guarantee that the ipcache shutdown has completed. However, we need to ensure that the shutdown completes within a reasonable time.

joestringer replied:
Broadly yeah that's what it's doing. The specifics here are relating to the way that events are enqueued below though, because if we *only* call `result.Trigger.Shutdown()` but then other packages continue to enqueue CIDRs into the asyncPrefixReleaser, then those calls will hang when they attempt to do the Trigger(). So this is wrapping that to also have this extra channel to avoid hanging on the trigger.

I hadn't thought too hard about an ipcache.Close/Shutdown function yet, that sounds like a reasonable alternative solution to this part of the problem. I'll look into that.

joestringer replied:
I looked a little closer, unfortunately it's not quite this simple. The reason that this doesn't confirm that the shutdown is complete is because the `pkg/trigger:Trigger.Shutdown()` method only closes a channel, so that function itself does not ensure the logic underneath completes. We'd need to make pkg/trigger more complicated to ensure that occurs synchronously. Hmm.

joestringer replied:
Maybe I can just go ahead with this structure anyway for now, then we can always further close that gap in future. I think it'll likely help the ongoing modularization effort if the IPCache has a shutdown function like this to handle cleanup.

aditighag replied:
>The reason that this doesn't confirm that the shutdown is complete is because the pkg/trigger:Trigger.Shutdown() method only closes a channel

Can we still move the logic to a shutdown function, and invoke that from daemon shutdown? In future, we can make the ipcache/gc.shutdown synchronous, and thereby guarantee a total shutdown (within bounded time). 

```
:
result.Trigger.Shutdown()
close(result.closeChan)
```

joestringer replied:
Almost yes - it turns out that the `Trigger` has a `ShutdownFunc()` which is called once the trigger completes its current task. Using that we can close the loop by having a shutdown function called from the daemon which calls Trigger.Shutdown(), which asynchronously stops any current trigger run then calls the Trigger.ShutdownFunc. From there we can `close(result.closeChan)`. Then, back in the overall ipcache shutdown func which called the `Trigger.Shutdown()` it can block on the close of `resultChan`, thereby ensuring that the trigger func is no longer executing any subsequent logic.

joestringer replied:
I've implemented this suggestion now. I'll mark the PR is draft until I see confirmation that I didn't break the tests then re-mark for review.

joestringer replied:
Thanks for the feedback!

aditighag commented on pkg/ipcache/gc.go:
```diff
@@ -41,15 +43,22 @@ func newAsyncPrefixReleaser(parent prefixReleaser, interval time.Duration) *asyn
-			//       from the Daemon.
-			ctx, cancel := context.WithTimeout(
-				context.TODO(),
+			innerCtx, cancel := context.WithTimeout(
```
I'm a bit wary of carrying multiple contexts. Although the inner context is derived from the daemon context, see the comment below. If that works we need not have multiple contexts.

joestringer replied:
In theory all contexts should derive from the top of the context tree so that you can cancel the top one and cancel all logic where that context or subcontexts are used. The way it didn't derive from the daemon before was weird, since it meant that the child context could outlive the overall agent. I can understand some reluctance around 'multiple contexts' if some code is trying to force two different parent contexts, since this isn't really handled well in the standard library. However, this is a pretty standard way to derive a context from a parent context. Maybe the only unusual thing here is that usually in the codebase we'd call one 'parentCtx' and one 'ctx'.

joestringer replied:
^ strictly speaking we could avoid doing the actual shutdown with this context, and instead do an explicit shutdown.. but even with that I think that this context should be derived from the agent context. So presume there's some logic going on using this context, first the agent context gets cancelled and we force this logic to cancel & return to caller. This also means that all subsequent calls with the ctx will fail quickly. Then separately the agent calls the explicit Shutdown() to ensure the goroutines stop etc. and nothing remains running.

aditighag reviewed (commented):
The changes look reasonable, but I have given an alternate suggestion on triggering the ipcache shutdown.

aditighag commented on pkg/ipcache/gc.go:
```diff
@@ -58,6 +67,16 @@ func newAsyncPrefixReleaser(parent prefixReleaser, interval time.Duration) *asyn
 	pr.Lock()
 	defer pr.Unlock()
+	select {
+	case <-pr.closeChan:
```
Can we block on the `result.Trigger channel` here?

joestringer replied:
We can't block directly on the result's trigger channel because it's not exposed. I've done the next best thing, PTAL.

joestringer committed:
ipcache: Plumb daemon context through IPCache

Plumb the primary daemon context into the label injector logic and the
async prefix releaser in order to reduce the likelihood of issues where
this logic continues to execute beyond the lifetime of the agent.

Found during unit testing, where the identity allocator would be closed
while the ipcache's garbage collection of identities was still running.

Suggested-by: Jussi Maki <jussi@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

aditighag commented on pkg/ipcache/ipcache.go:
```diff
@@ -122,10 +124,20 @@ func NewIPCache(c *Configuration) *IPCache {
 	}
-	ipc.deferredPrefixRelease = newAsyncPrefixReleaser(ipc, 1*time.Millisecond)
+	ctx := context.Background()
+	if c != nil && c.Context != nil {
+		ctx = c.Context
```
I suppose the context `nil` check is mainly added for unit tests? Can we not pass the `config.Context` in unit tests as well for consistency? This will make it clear that the ipcache module expects a context to be passed, and that the internal logic checks if this context is done (I haven't checked this part though).

joestringer replied:
I must admit, I was avoiding this mainly because I knew there was a bit of tech debt there. But as it turns out, we can do some clean additional separation of packages / components by following this to the logical conclusion.

I'll open a follow up PR to review the solution to this since it spreads across a whole bunch of packages and the issue being fixed in this PR is currently affecting the master branch.

joestringer replied:
See https://github.com/cilium/cilium/pull/21774 for the follow up.

aditighag reviewed (approved):
The shutdown logic looks much more deterministic now. Thanks for the revisions.
I've posted a question about the plumbed context, but the fix looks good to me.

Edit: The travis failure is in a clustermesh test that's probably unrelated to the changes. Mainly, the ipcache/gc* unit test has passed. Do we need to trigger the Travis test couple of times?
